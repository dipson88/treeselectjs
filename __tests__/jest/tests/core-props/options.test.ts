import { fireEvent } from '@testing-library/dom'
import {
  classes,
  getArrowElement,
  getListGroupsItems,
  getListItems,
  getTagsElement,
  noResultsText,
  renderTreeselect
} from '../../helpers'

const { list: listClasses } = classes

describe('options prop', () => {
  it('should render a Treeselect with empty options', () => {
    const treeselect = renderTreeselect({
      value: [],
      options: []
    })

    const arrow = getArrowElement(treeselect.parentHtmlContainer)
    fireEvent.mouseDown(arrow)

    expect(document.body.innerHTML).toContain(noResultsText)
    expect(treeselect.options).toEqual([])
  })

  it('should render a Treeselect with options', () => {
    const treeselect = renderTreeselect({
      value: [],
      options: [
        { value: 1, name: 'Option 1', children: [] },
        { value: 2, name: 'Option 2', children: [] },
        { value: 3, name: 'Option 3', children: [] }
      ]
    })

    const arrow = getArrowElement(treeselect.parentHtmlContainer)
    fireEvent.mouseDown(arrow)
    const options = getListItems(treeselect.parentHtmlContainer)

    expect(options.length).toBe(3)
    expect(treeselect.options.length).toEqual(3)
  })

  it('should contain attributes in options', () => {
    const treeselect = renderTreeselect({
      value: [],
      options: [
        { value: 1, name: 'Option 1', children: [], htmlAttr: { test: 'test-data-1' } },
        { value: 2, name: 'Option 2', children: [] },
        { value: 3, name: 'Option 3', children: [] }
      ]
    })

    const arrow = getArrowElement(treeselect.parentHtmlContainer)
    fireEvent.mouseDown(arrow)
    const firstOption = getListItems(treeselect.parentHtmlContainer)[0]

    expect(firstOption.getAttribute('test')).toBe('test-data-1')
    expect(firstOption.innerHTML).toContain('Option 1')
  })

  it('should contain attributes in children options', () => {
    const treeselect = renderTreeselect({
      value: [],
      options: [
        {
          value: 1,
          name: 'Option 1',
          children: [
            { value: 2, name: 'Option 2', children: [], htmlAttr: { test: 'test-data-2' } },
            { value: 3, name: 'Option 3', children: [] }
          ]
        }
      ]
    })

    const arrow = getArrowElement(treeselect.parentHtmlContainer)
    fireEvent.mouseDown(arrow)
    const options = getListItems(treeselect.parentHtmlContainer)
    const secondOption = options[1]

    expect(secondOption.getAttribute('test')).toBe('test-data-2')
    expect(secondOption.innerHTML).toContain('Option 2')
  })

  it('should contain disabled class in options', () => {
    const treeselect = renderTreeselect({
      value: [],
      options: [
        { value: 1, name: 'Option 1', children: [], disabled: true },
        { value: 2, name: 'Option 2', children: [] },
        { value: 3, name: 'Option 3', children: [] }
      ]
    })

    const arrow = getArrowElement(treeselect.parentHtmlContainer)
    fireEvent.mouseDown(arrow)
    const firstOption = getListItems(treeselect.parentHtmlContainer)[0]

    expect(firstOption.classList.contains(listClasses.itemDisabled)).toBe(true)
  })

  it('should not select disabled option', () => {
    const treeselect = renderTreeselect({
      value: [],
      options: [
        { value: 1, name: 'Option 1', children: [], disabled: true },
        { value: 2, name: 'Option 2', children: [] },
        { value: 3, name: 'Option 3', children: [] }
      ]
    })

    const arrow = getArrowElement(treeselect.parentHtmlContainer)
    fireEvent.mouseDown(arrow)
    const firstOption = getListItems(treeselect.parentHtmlContainer)[0]
    fireEvent.mouseDown(firstOption)
    const tagsElement = getTagsElement(treeselect.parentHtmlContainer)

    expect(treeselect.value).toEqual([])
    expect(tagsElement.innerHTML).not.toContain('Option 1')
  })

  it('should render a Treeselect with children options', () => {
    const treeselect = renderTreeselect({
      value: [],
      options: [
        {
          value: 1,
          name: 'Option 1',
          children: [
            { value: 2, name: 'Option 2', children: [] },
            { value: 3, name: 'Option 3', children: [] }
          ]
        }
      ]
    })

    const arrow = getArrowElement(treeselect.parentHtmlContainer)
    fireEvent.mouseDown(arrow)
    const options = getListItems(treeselect.parentHtmlContainer)

    expect(options.length).toBe(3)
    expect(treeselect.options.length).toEqual(1)
    expect(treeselect.options[0].children.length).toEqual(2)
  })

  it('should not select disabled children option', () => {
    const treeselect = renderTreeselect({
      value: [],
      options: [
        {
          value: 1,
          name: 'Option 1',
          children: [
            { value: 2, name: 'Option 2', children: [], disabled: true },
            { value: 3, name: 'Option 3', children: [] }
          ]
        }
      ]
    })

    const arrow = getArrowElement(treeselect.parentHtmlContainer)
    fireEvent.mouseDown(arrow)
    const options = getListItems(treeselect.parentHtmlContainer)
    const secondOption = options[1]
    fireEvent.mouseDown(secondOption)
    const tagsElement = getTagsElement(treeselect.parentHtmlContainer)

    expect(treeselect.value).toEqual([])
    expect(tagsElement.innerHTML).not.toContain('Option 2')
  })

  it('should select children if all children are selected', () => {
    const treeselect = renderTreeselect({
      value: [],
      options: [
        {
          value: 1,
          name: 'Option 1',
          children: [
            { value: 2, name: 'Option 2', children: [] },
            { value: 3, name: 'Option 3', children: [] }
          ]
        }
      ]
    })

    const arrow = getArrowElement(treeselect.parentHtmlContainer)
    fireEvent.mouseDown(arrow)
    const options = getListItems(treeselect.parentHtmlContainer)
    const secondOption = options[1]
    const thirdOption = options[2]
    fireEvent.mouseDown(secondOption)
    fireEvent.mouseDown(thirdOption)
    const tagsElement = getTagsElement(treeselect.parentHtmlContainer)

    expect(treeselect.value).toEqual([2, 3])
    expect(tagsElement.innerHTML).toContain('Option 1')
  })

  it('should select non disabled children if all children are selected and one is disabled', () => {
    const treeselect = renderTreeselect({
      value: [],
      options: [
        {
          value: 1,
          name: 'Option 1',
          children: [
            { value: 2, name: 'Option 2', children: [], disabled: true },
            { value: 3, name: 'Option 3', children: [] }
          ]
        }
      ]
    })

    const arrow = getArrowElement(treeselect.parentHtmlContainer)
    fireEvent.mouseDown(arrow)
    const options = getListItems(treeselect.parentHtmlContainer)
    const thirdOption = options[2]
    fireEvent.mouseDown(thirdOption)
    const tagsElement = getTagsElement(treeselect.parentHtmlContainer)

    expect(treeselect.value).toEqual([3])
    expect(tagsElement.innerHTML).toContain('Option 3')
  })

  it('should make groups unselectable if option contains isGroupSelectable: false', () => {
    const disabledGroupName = 'Option 2'
    const disabledGroupValue = 2
    const disabledItemName = 'Option 3'
    const disabledItemValue = 3

    const treeselect = renderTreeselect({
      value: [],
      options: [
        {
          value: 1,
          name: 'Option 1',
          children: [
            {
              value: disabledGroupValue,
              name: disabledGroupName,
              isGroupSelectable: false,
              children: [
                {
                  value: 4,
                  name: 'Option 4',
                  children: []
                }
              ]
            },
            {
              value: disabledItemValue,
              name: disabledItemName,
              isGroupSelectable: false,
              children: []
            }
          ]
        }
      ]
    })

    treeselect.toggleOpenClose()

    // Checked isGroupSelectable for group item
    const [groupItem] = Array.from(getListGroupsItems(treeselect.parentHtmlContainer)).filter(
      (item) => item.getAttribute('title') === disabledGroupName
    )

    fireEvent.mouseDown(groupItem)

    expect(treeselect.value).toEqual([])
    expect(groupItem.classList.contains(listClasses.itemNonSelectableGroup)).toBe(true)

    // Checked isGroupSelectable for regular item
    const [regularItem] = Array.from(getListItems(treeselect.parentHtmlContainer)).filter(
      (item) => item.getAttribute('title') === disabledItemName
    )

    fireEvent.mouseDown(regularItem)

    expect(treeselect.value).toEqual([disabledItemValue])
    expect(regularItem.classList.contains(listClasses.itemNonSelectableGroup)).toBe(false)

    // Checked html structure
    expect(treeselect.parentHtmlContainer).toMatchSnapshot()
  })
})
