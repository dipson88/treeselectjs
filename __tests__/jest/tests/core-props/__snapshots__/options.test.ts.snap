// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`options prop should make groups unselectable if option contains isGroupSelectable: false 1`] = `
<div
  class="treeselect"
  id="test-container"
>
  <div
    class="treeselect-input treeselect-input--bottom treeselect-input--opened"
    tabindex="-1"
  >
    <div
      class="treeselect-input__tags"
    >
      <div
        class="treeselect-input__tags-element"
        tabindex="-1"
        tag-id="3"
        title="Option 3"
      >
        <span
          class="treeselect-input__tags-name"
        >
          Option 3
        </span>
        <span
          class="treeselect-input__tags-cross"
        >
          <svg
            fill="none"
            height="15"
            stroke="#000000"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 25 25"
            width="15"
            xmlns="http://www.w3.org/2000/svg"
          >
            <line
              x1="18"
              x2="6"
              y1="6"
              y2="18"
            />
            <line
              x1="6"
              x2="18"
              y1="6"
              y2="18"
            />
          </svg>
        </span>
      </div>
      <input
        class="treeselect-input__edit"
      />
    </div>
    <div
      class="treeselect-input__operators"
    >
      <span
        class="treeselect-input__clear"
        tabindex="-1"
      >
        <svg
          fill="none"
          height="15"
          stroke="#000000"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 25 25"
          width="15"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="10"
          />
          <line
            x1="15"
            x2="9"
            y1="9"
            y2="15"
          />
          <line
            x1="9"
            x2="15"
            y1="9"
            y2="15"
          />
        </svg>
      </span>
      <span
        class="treeselect-input__arrow"
      >
        <svg
          fill="none"
          height="15"
          stroke="#000000"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 25 25"
          width="15"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M18 15l-6-6-6 6"
          />
        </svg>
      </span>
    </div>
  </div>
  <div
    class="treeselect-list treeselect-list--bottom"
    direction="bottom"
  >
    <div
      class="treeselect-list__group-container"
      group-container-id="1"
    >
      <div
        class="treeselect-list__item treeselect-list__item--group treeselect-list__item--closed treeselect-list__item--focused treeselect-list__item--partial-checked"
        group="true"
        level="0"
        style="padding-left: 0px;"
        tabindex="-1"
        title="Option 1"
      >
        <span
          class="treeselect-list__item-icon"
          tabindex="-1"
        >
          <svg
            fill="none"
            height="15"
            stroke="#000000"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 25 25"
            width="15"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9 18l6-6-6-6"
            />
          </svg>
        </span>
        <div
          class="treeselect-list__item-checkbox-container"
        >
          <span
            class="treeselect-list__item-checkbox-icon"
          >
            <svg
              fill="none"
              height="15"
              stroke="#000000"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 25 25"
              width="15"
              xmlns="http://www.w3.org/2000/svg"
            >
              <line
                x1="5"
                x2="19"
                y1="12"
                y2="12"
              />
            </svg>
          </span>
          <input
            class="treeselect-list__item-checkbox"
            input-id="1"
            tabindex="-1"
            type="checkbox"
          />
        </div>
        <label
          class="treeselect-list__item-label"
        >
          Option 1
        </label>
      </div>
      <div
        class="treeselect-list__group-container"
        group-container-id="2"
      >
        <div
          class="treeselect-list__item treeselect-list__item--group treeselect-list__item--hidden treeselect-list__item--non-selectable-group"
          group="true"
          level="1"
          style="padding-left: 20px;"
          tabindex="-1"
          title="Option 2"
        >
          <span
            class="treeselect-list__item-icon"
            tabindex="-1"
          >
            <svg
              fill="none"
              height="15"
              stroke="#000000"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 25 25"
              width="15"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M6 9l6 6 6-6"
              />
            </svg>
          </span>
          <div
            class="treeselect-list__item-checkbox-container"
          >
            <span
              class="treeselect-list__item-checkbox-icon"
            />
            <input
              class="treeselect-list__item-checkbox"
              input-id="2"
              tabindex="-1"
              type="checkbox"
            />
          </div>
          <label
            class="treeselect-list__item-label"
          >
            Option 2
          </label>
        </div>
        <div
          class="treeselect-list__item"
          group="false"
          level="2"
          style="padding-left: 60px;"
          tabindex="-1"
          title="Option 4"
        >
          <div
            class="treeselect-list__item-checkbox-container"
          >
            <span
              class="treeselect-list__item-checkbox-icon"
            />
            <input
              class="treeselect-list__item-checkbox"
              input-id="4"
              tabindex="-1"
              type="checkbox"
            />
          </div>
          <label
            class="treeselect-list__item-label"
          >
            Option 4
          </label>
        </div>
      </div>
      <div
        class="treeselect-list__item treeselect-list__item--hidden treeselect-list__item--checked"
        group="false"
        level="1"
        style="padding-left: 40px;"
        tabindex="-1"
        title="Option 3"
      >
        <div
          class="treeselect-list__item-checkbox-container"
        >
          <span
            class="treeselect-list__item-checkbox-icon"
          >
            <svg
              fill="none"
              height="15"
              stroke="#000000"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 25 25"
              width="15"
              xmlns="http://www.w3.org/2000/svg"
            >
              <polyline
                points="20 6 9 17 4 12"
              />
            </svg>
          </span>
          <input
            class="treeselect-list__item-checkbox"
            input-id="3"
            tabindex="-1"
            type="checkbox"
          />
        </div>
        <label
          class="treeselect-list__item-label"
        >
          Option 3
        </label>
      </div>
    </div>
    <div
      class="treeselect-list__empty treeselect-list__empty--hidden"
      title="No results found..."
    >
      <span
        class="treeselect-list__empty-icon"
      >
        <svg
          fill="none"
          height="15"
          stroke="#000000"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 25 25"
          width="15"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10.29 3.86L1.82 18a2 2 0 0 0 1.71 3h16.94a2 2 0 0 0 1.71-3L13.71 3.86a2 2 0 0 0-3.42 0z"
          />
          <line
            x1="12"
            x2="12"
            y1="9"
            y2="13"
          />
          <line
            x1="12"
            x2="12.01"
            y1="17"
            y2="17"
          />
        </svg>
      </span>
      <span
        class="treeselect-list__empty-text"
      >
        No results found...
      </span>
    </div>
  </div>
</div>
`;
